src/data/pfun/fix.lean

```diff
@@ -0,0 +1,319 @@
+    split; intro hh, exact ⟨_,hh⟩,
+    have h₁ := p_find h₀,
+    rw [dom_iff_mem] at h₁,
+    cases h₁ with y h₁,
+    replace h₁ := approx_mono' hf _ _ h₁,
+    suffices : y = b, subst this, exact h₁,
+    cases hh with i hh,
+    revert h₁, generalize : (succ (find h₀)) = j, intro,
+    wlog : i ≤ j := le_total i j using [i j b y,j i y b],
+    replace hh := approx_mono hf case _ _ hh,
+    apply roption.mem_unique h₁ hh },
+  { simp [fix_def' f h₀],
+    simp [dom_iff_mem] at h₀,
+    intro, apply h₀ }
+end
+
+lemma max_fix (i : ℕ) : approx f i ≤ fix f :=
+assume a b hh,
+by { rw [mem_fix hf], exact ⟨_,hh⟩ }
+
+lemma min_fix (x : α) : ∃ i, fix f x ≤ approx f i x :=
+begin
+  by_cases hh : ∃ i b, b ∈ approx f i x,
+  { rcases hh with ⟨i,b,hb⟩, existsi i,
+    intros b' h',
+    have hb' := max_fix hf i _ _ hb,
+    have hh := roption.mem_unique h' hb',
+    subst hh, exact hb },
+  { simp at hh, existsi 0,
+    intros b' h', simp [mem_fix hf] at h',
+    cases h' with i h',
+    cases hh _ _ h' }
+end
+
+-- lemma approx_eq {i j : ℕ} {a : α} (hi : (approx f i a).dom) (hj : (approx f j a).dom) :
+--   approx f i a = approx f j a :=
+-- begin
+--   simp [dom_iff_mem] at hi hj,
+--   cases hi with b hi, cases hj with b' hj,
+--   have : b' = b,
+--   { wlog hij : i ≤ j := le_total i j using [i j b b',j i b' b],
+--     apply mem_unique hj, apply approx_mono hf hij _ _ hi, },
+--   subst b',
+--   ext y, split; intro hy,
+--   rename hi hh, rename hj hi, rename hh hj,
+--   all_goals
+--   { have := mem_unique hy hj, subst this,
+--     exact hi }
+-- end
+
+noncomputable def approx_chain : chain (Π a, roption $ β a) :=
+begin
+  refine ⟨ approx f, _ ⟩,
+  apply approx_mono, exact hf
+end
+
+lemma le_f_of_mem_approx {x} (hx : x ∈ approx_chain hf) : x ≤ f x :=
+begin
+  revert hx, simp [approx_chain,stream.mem_def],
+  intros i hx, subst x,
+  apply approx_mono' hf
+end
+
+lemma f_mem_approx_chain {x} (hx : x ∈ approx_chain hf) : f x ∈ approx_chain hf :=
+begin
+  revert hx, simp [approx_chain,stream.mem_def],
+  intros i hx, subst hx, exact ⟨i.succ,rfl⟩
+end
+
+lemma approx_mem_approx_chain {i} : approx f i ∈ approx_chain hf :=
+stream.mem_of_nth_eq rfl
+
+end fix
+
+namespace pi
+
+open fix
+
+variables {α} {β : α → Type*}
+open complete_partial_order
+
+variables {f : (Π a, roption $ β a) → (Π a, roption $ β a)} (hf : monotone f)
+
+include hf
+
+open roption (hiding Sup) nat
+open nat.up complete_partial_order
+
+lemma fix_eq_Sup : fix f = Sup (approx_chain hf) :=
+begin
+  apply le_antisymm,
+  { intro x, cases min_fix hf x with i hx,
+    transitivity' approx f i.succ x,
+    { transitivity', apply hx, apply approx_mono' hf },
+    apply le_Sup _ _ _, dsimp [approx],
+    rw chain.mem_map_iff,
+    refine ⟨approx f i.succ,_,rfl⟩,
+    apply approx_mem_approx_chain },
+  { apply Sup_le _ _ _,
+    simp [mem_map_iff,approx_chain,stream.mem_def],
+    intros y x, revert y, simp, apply max_fix hf },
+end
+
+@[main_declaration]
```

Comment:
This looks mysterious...

Reply:
How so?

Reply:
It's an attribute I haven't seen before, with a nondescript name. What tactic is it associated with?

Reply:
It is indeed brand new: #3701

Reply:
I see, I missed the bikeshedding. Rob also mentions in that PR that there is an expectation that mathlib PRs should not have `@[needed]` in them, and the commented code in this PR should also be removed. And of course more docs, always more docs.

Reply:
More docs, indeed. It's coming. I think removing those tags right before merging can help the review. It shows you what is there to support and what is a feature of the PR.